File: 프로그래머스/2/87946. 피로도/피로도.swift
import Foundation

func solution(_ k:Int, _ dungeons:[[Int]]) -> Int {
    
    var visited = Array(repeating: false, count: dungeons.count)
    var selected = Array(repeating: [0, 0], count: dungeons.count)
    var maxCount = 0
    
    
    func calculateCount() -> Int {
        var tempFirst = k
        var count = 0
        for select in selected {
            if tempFirst >= select[0] {
                tempFirst = tempFirst - select[1]
                count += 1
            }
        }
        return count
    }
    
    func DFS(count: Int) {
        if count == dungeons.count {
            let tempCount = calculateCount()
            if maxCount < tempCount {
                maxCount = tempCount
            }
            return
        }

        for i in 0..<dungeons.count { // 0, 1, 2
            if visited[i] == false {
                visited[i] = true
                selected[count] = dungeons[i]
                DFS(count: count+1)
                visited[i] = false
            }
        }
    }

    DFS(count: 0)
    return maxCount
}

File: 프로그래머스/2/87946. 피로도/피로도.py
def solution(k, dungeons):
    answer = 0
    length = len(dungeons)
    visited = [False for _ in range(length)]
    
    def DFS(blood, number):
        nonlocal answer
        answer = max(number, answer)
        for i in range(length):
            if not visited[i]:
                if blood >= dungeons[i][0] and blood - dungeons[i][1] >= 0:
                    visited[i] = True
                    DFS(blood-dungeons[i][1], number+1)
                    visited[i] = False
    DFS(k, 0)
    return answer